### DISCUSSION TITLE
AuthStore for React Native

### DISCUSSION THREAD

**Discussion by lucaslevin** [pocketbase/pocketbase#1856] [Q&A] [answered]
Hello all. I am in need of writing an AuthStore that use AsyncStorage instead of localStorage but I am having a rough time figuring it out. Hopefully someone can help me out as I would like to use PocketBase in React Native.

**Comment by ganigeorgiev** [comment, marked as answer]
**Edit: The latest JS SDK comes with builtin AsyncAuthStore out of the box - https://github.com/pocketbase/js-sdk#asyncauthstore.**

---

As mentioned in https://github.com/pocketbase/js-sdk/issues/142, the `AuthStore` by design is synchronous and it will not work out of the box with async implementations.

One way to incorporate an async store could be to maintain some sort of queue that will serialize the store changes to prevent race conditions. You can use an external library for that or implement a small helper on your own (_technically you need only to be able to "push" and "process" the queue_).

Here is a simple plain js example (**I haven't tested it and some adjustments may be necessary**):

```js
import { BaseAuthStore } from "pocketbase";
import AsyncStorage from "..." // replace with the async storage library you use

class AsyncAuthStore extends BaseAuthStore {
    /**
     * @param {String} storageKey
     */
    constructor(storageKey = "pb_auth") {
        super();

        this.storageKey = storageKey;
        this.queue = [];

        this._enqueue(async () => {
            const raw = await AsyncStorage.getItem(this.storageKey);
            if (raw) {
                const decoded = JSON.parse(raw);
                this.save(decoded.token, decoded.model);
            }
        })
    }

    /**
     * @inheritdoc
     */
    save(token, model) {
        super.save(token, model);
    
        this._enqueue(() => {
            return AsyncStorage.setItem(this.storageKey, JSON.stringify({ token, model }))
        })
    }

    /**
     * @inheritdoc
     */
    clear() {
        super.clear();

        this._enqueue(() => {
            return AsyncStorage.removeItem(this.storageKey)
        })
    }

    /**
     * Appends an async function to the queue.
     */
    _enqueue(asyncCallback) {
        this.queue.push(asyncCallback);

        if (this.queue.length == 1) {
            _dequeue();
        }
    }

    /**
     * Starts the queue processing.
     */
    _dequeue() {
        if (!this.queue.length) {
            return;
        }

        this.queue[0]().finally(() => {
            this.queue.shift();

            if (!this.queue.length) {
                return;
            }

            _dequeue();
        });
    }
}
```

And then to use the store, you have to pass it as constructor argument during the SDK client initialization:

```js
import PocketBase from "pocketbase";
import AsyncAuthStore from "./AsyncAuthStore"

const pb = new PocketBase("http://127.0.0.1:8090", new AsyncAuthStore());
```

Some notes:

- Most of the above code is the queue handling. If you don't like it, you can look for a more complete and tested sync queue library.
- The above maintain the data in 2 places - in runtime/memory (the original `BaseAuthStore` sync behavior) and in the async store.

**Reply by lucaslevin** [reply]
That works pretty well. I will test some more, might come back if I have any findings. Thank you so much and sorry for the inconvenience!
